File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/data/KStore.java
package org.kevoree.modeling.api.data;

import org.kevoree.modeling.api.Callback;
import org.kevoree.modeling.api.KUniverse;
import org.kevoree.modeling.api.KObject;
import org.kevoree.modeling.api.KView;
import org.kevoree.modeling.api.event.KEventBroker;
import org.kevoree.modeling.api.util.KOperationManager;

/**
 * Created by duke on 10/17/14.
 */
public interface KStore {

    void lookup(KView originView, Long key, Callback<KObject> callback);

    void lookupAll(KView originView, Long[] key, Callback<KObject[]> callback);

    Object[] raw(KObject origin, AccessMode accessMode);

    public void save(KUniverse universe, Callback<Throwable> callback);

    public void saveUnload(KUniverse universe, Callback<Throwable> callback);

    public void discard(KUniverse universe, Callback<Throwable> callback);

    public void delete(KUniverse universe, Callback<Throwable> callback);

    public void initKObject(KObject obj, KView originView);

    public void initUniverse(KUniverse universe);

    long nextUniverseKey();

    long nextObjectKey();

    public void getRoot(KView originView, Callback<KObject> callback);

    public void setRoot(KObject newRoot, Callback<Throwable> callback);

    public KEventBroker eventBroker();

    public void setEventBroker(KEventBroker broker);

    public KDataBase dataBase();

    public void setDataBase(KDataBase dataBase);

    public KOperationManager operationManager();

    public void connect(Callback<Throwable> callback);

    public void close(Callback<Throwable> callback);

}

File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/data/KStore.ts
///<reference path="../Callback.ts"/>
///<reference path="../KDimension.ts"/>
///<reference path="../KEvent.ts"/>
///<reference path="../KObject.ts"/>
///<reference path="../KView.ts"/>
///<reference path="../ModelListener.ts"/>
///<reference path="../time/TimeTree.ts"/>
///<reference path="../../../../../java/util/List.ts"/>
///<reference path="../../../../../java/util/Set.ts"/>

interface KStore {

  lookup(originView: KView, key: number, callback: Callback<KObject<any,any>>): void;

  lookupAll(originView: KView, key: Set<number>, callback: Callback<List<KObject<any,any>>>): void;

  raw(origin: KObject<any,any>, accessMode: AccessMode): any[];

  save(dimension: KDimension<any,any,any>, callback: Callback<Throwable>): void;

  saveUnload(dimension: KDimension<any,any,any>, callback: Callback<Throwable>): void;

  discard(dimension: KDimension<any,any,any>, callback: Callback<Throwable>): void;

  delete(dimension: KDimension<any,any,any>, callback: Callback<Throwable>): void;

  timeTree(dimension: KDimension<any,any,any>, key: number, callback: Callback<TimeTree>): void;

  timeTrees(dimension: KDimension<any,any,any>, keys: number[], callback: Callback<TimeTree[]>): void;

  initKObject(obj: KObject<any,any>, originView: KView): void;

  initDimension(dimension: KDimension<any,any,any>, callback: Callback<Throwable>): void;

  nextDimensionKey(): number;

  nextObjectKey(): number;

  getDimension(key: number): KDimension<any,any,any>;

  getRoot(originView: KView, callback: Callback<KObject<any,any>>): void;

  setRoot(newRoot: KObject<any,any>): void;

  cacheLookup(dimension: KDimension<any,any,any>, time: number, key: number): KObject<any,any>;

  registerListener(origin: any, listener: ModelListener): void;

  notify(event: KEvent): void;

}

